### DISCUSSION TITLE
How can I invalidate dependent providers?

### DISCUSSION THREAD

**Discussion by hisaichi5518** [rrousselGit/riverpod#3743] [Q&A]
I'm using Riverpod and have Provider A and Provider B. Provider B depends on Provider A. A Widget depends on Provider B, but the Widget doesn't know that Provider B depends on Provider A. When the Widget wants to recreate Provider B's data from scratch, it invalidates Provider B. However, since Provider B depends on Provider A, there's no change in the value.

```dart
// provider.dart
final providerA = Provider((ref) => ...);

final providerB = Provider((ref) {
  final a = ref.watch(providerA);
  return a * 2;
});

// widget.dart
class MyWidget ... {
  Widget build(context, ref) {
    return TextButton(onPressed: () {
      // Invalidating providerB, but since it doesn't invalidate providerA itself, there's no change in value
      ref.invalidate(providerB);
    });
  }
}
```

To change the value, we need to invalidate providerA. However, I don't think the Widget should know the implementation detail of Provider B, which is "Provider B depends on Provider A". Is there a solution for this?

EDIT:
By using `ref.onDispose` to perform invalidation, we can encapsulate the invalidation of providerA within providerB. However, when multiple Providers use onDispose to invalidate providerA, it appears that the invalidation of providerA is executed multiple times.

```dart
// provider.dart
final providerA = Provider((ref) => ...);

final providerB = Provider((ref) {
  ref.onDispose((ref) {
    ref.invalidate(providerA);
  });
  final a = ref.watch(providerA);
  return a * 2;
});

final providerC = Provider((ref) {
  ref.onDispose((ref) {
    // If providerB is invalidated, providerA will be invalidated and providerC will be recreated
    // Being recreated means onDispose will be executed, and invalidate(providerA) will be called
    ref.invalidate(providerA);
  });
  final a = ref.watch(providerA);
  return a * 3;
});
```

**Comment by gdurandrexel** [comment]
You coud transforn B in a NotifierProvider and add a refresh() method that invalidates A that would be called by your widget

**Reply by hisaichi5518** [reply]
Indeed, it seems that by using a Notifier, we can avoid the need to know about invalidating provider A.
However, the fact that we need to use `ref.read(providerBProvider.notifier).refresh()` instead of `ref.invalidate()` provided by Riverpod means we still need to know about the implementation of providerB.

@rrousselGit 
Does Riverpod have any plans to provide a method to invalidate a provider along with the providers it depends on?
